staging/src/k8s.io/apiserver/pkg/server/options/admission.go

```diff
@@ -146,32 +143,39 @@ func (a *AdmissionOptions) Validate() []error {
 	return errs
 }
 
-// enabledPluginNames makes use of RecommendedPluginOrder and DefaultOffPlugins fields
+// enabledPluginNames makes use of RecommendedPluginOrder, DefaultEnabledPlugins and PluginNames fields
 // to prepare a list of plugin names that are enabled.
-//
-// TODO(p0lyn0mial): In the end we will introduce two new flags:
-// --disable-admission-plugin this would be a list of admission plugins that a cluster-admin wants to explicitly disable.
-// --enable-admission-plugin  this would be a list of admission plugins that a cluster-admin wants to explicitly enable.
-// both flags are going to be handled by this method
 func (a *AdmissionOptions) enabledPluginNames() []string {
-	//TODO(p0lyn0mial): first subtract plugins that a user wants to explicitly enable from allOffPlugins (DefaultOffPlugins)
-	//TODO(p0lyn0miial): then add/append plugins that a user wants to explicitly disable to allOffPlugins
-	//TODO(p0lyn0mial): so that --off=three --on=one,three default-off=one,two results in  "one" being enabled.
-	allOffPlugins := a.DefaultOffPlugins
-	onlyEnabledPluginNames := []string{}
-	for _, pluginName := range a.RecommendedPluginOrder {
-		disablePlugin := false
-		for _, disabledPluginName := range allOffPlugins {
-			if pluginName == disabledPluginName {
-				disablePlugin = true
-				break
-			}
+	enabledPlugins := []string{}
+	allPlugins := []string{}
+	removeDuplicated := sets.NewString()
+
+	// 1.remove duplicated both in PluginNames and DefaultEnabledPlugins
+	for _, plugin := range append(a.PluginNames, a.DefaultEnabledPlugins...) {
+		if !removeDuplicated.Has(plugin) {
+			removeDuplicated.Insert(plugin)
+			allPlugins = append(allPlugins, plugin)
+		}
+	}
+
+	enabledPluginsSet := sets.NewString(allPlugins...)
+	recommendedOrderedPluginsSet := sets.NewString(a.RecommendedPluginOrder...)
+
+	// 2.get enabled plugins with recommended orders
+	orderedPluginsSet := enabledPluginsSet.Intersection(recommendedOrderedPluginsSet)
+	for _, recommenedPlugin := range a.RecommendedPluginOrder {
+		if orderedPluginsSet.Has(recommenedPlugin) {
+			enabledPlugins = append(enabledPlugins, recommenedPlugin)
 		}
-		if disablePlugin {
-			continue
+	}
+
+	// 3.get enabled plugins without recommended order, these ones are ordered by specified
+	differentSet := enabledPluginsSet.Difference(recommendedOrderedPluginsSet)
+	for _, plugin := range allPlugins {
+		if differentSet.Has(plugin) {
+			enabledPlugins = append(enabledPlugins, plugin)
```

Comment:
I tend to expect an error instead of appending them blindly. @deads2k opinion?

Reply:
Same as above https://github.com/kubernetes/kubernetes/pull/58123#discussion_r160893971